File: erlang/skynet.erl
-module(skynet).
-export([skynet/4, skynet/2]).
skynet(Parent, Num, 1, _) ->
  Parent ! Num;
skynet(Parent, Num, Size, Div) ->
  NewSize = Size div Div,
  lists:foreach(fun(X) -> spawn(skynet, skynet, [self(), Num + X * NewSize, NewSize, Div]) end, lists:seq(0, Div - 1)),
  process_sum(Parent, Div, 0, 0).

process_sum(Parent, Div, Received, Total) ->
  receive
    N ->
      if
        Received < Div - 1 -> process_sum(Parent, Div, Received + 1, Total + N);
        true -> Parent ! Total + N
      end
  end.

skynet(Size, Div) ->
  Start = erlang:system_time(milli_seconds),
  spawn(skynet, skynet, [self(), 0, Size, Div]),
  receive
    N ->
      End = erlang:system_time(milli_seconds),
      Diff = End - Start, 
      io:fwrite("Result: ~w in ~w ms.~n",[N, Diff])
  end.

File: scala/src/main/scala/Skynet.scala
import akka.actor.{ActorSystem, Props, ActorRef, Actor}

class Skynet(parent: ActorRef, num: Int, size: Int, div: Int) extends Actor {

  var received = 0
  var total = 0L

  size match {
    case 1 => parent ! num.toLong
    case x if x > 1 => (0 until div).map(mkChild)
  }

  override def receive = {
    case x: Long => received += 1; total += x; if (div == received) parent ! total
  }

  private def mkChild(n: Int): ActorRef = context.system.actorOf(Props(classOf[Skynet], self, num + n * (size / div), size / div, div))
}

class Root extends Actor {

  val startTime = System.currentTimeMillis()
  context.system.actorOf(Props(classOf[Skynet], self, 0, 1000000, 10))

  override def receive = {
    case x: Long =>
      val diffMs = System.currentTimeMillis() - startTime
      println(s"Result: $x in $diffMs ms.")
  }
}

object Root extends App {
  ActorSystem.create("main").actorOf(Props[Root])
}
